src/OrbitGl/TimerTrack.cpp

```diff
@@ -395,3 +412,39 @@ internal::DrawData TimerTrack::GetDrawData(
 size_t TimerTrack::GetNumberOfTimers() const { return timer_data_->GetNumberOfTimers(); }
 uint64_t TimerTrack::GetMinTime() const { return timer_data_->GetMinTime(); }
 uint64_t TimerTrack::GetMaxTime() const { return timer_data_->GetMaxTime(); }
+
+constexpr float kMinimalWidthToHaveBorder = 4.0;
+
+bool TimerTrack::ShouldHaveBorder(
+    const TimerInfo* timer, const std::optional<orbit_statistics::HistogramSelectionRange>& range,
+    float width) const {
+  if ((!range.has_value() || width < kMinimalWidthToHaveBorder || !app_->HasCaptureData()) ||
+      (app_->GetCaptureData().ProvideScopeId(*timer) != app_->GetHighlightedScopeId())) {
+    return false;
+  }
+  const uint64_t duration = timer->end() - timer->start();
+  return range->min_duration <= duration && duration <= range->max_duration;
+}
+
+[[nodiscard]] static Vec2 Vec3ToVec2(const Vec3& v) { return {v[0], v[1]}; }
+
+void TimerTrack::AddBorderLine(const Vec2& from, const Vec2& to, float z, const Color& color,
+                               PrimitiveAssembler& primitive_assembler,
+                               const orbit_client_protos::TimerInfo& timer_info) {
+  auto user_data = CreatePickingUserData(primitive_assembler, timer_info);
+  primitive_assembler.AddLine(from, to, z, color, std::move(user_data));
+}
+
+void TimerTrack::AddTetragonBorder(PrimitiveAssembler& primitive_assembler,
```

Comment:
As I said above, this method is a clear candidate to be in the new PrimitiveAssembler (and old Batcher) class.